### DISCUSSION TITLE
podman login successful for invalid registries

### DISCUSSION THREAD

**Discussion by gshepherd7** [containers/podman#12988] [Q&A]
Noticed this as a result of a typo, but it appears that when doing a podman login, as long as a curl GET <registry> returns success, that podman will report successful login regardless of the credentials entered.  My example...

`curl https://test2.di2e.net/v2/ `

302 Found

`sudo podman login https://test2.di2e.net/v2/`

Username: asdfadf
Password: 
Login Succeeded!

A registry by this name does not exist, however the curl command is reporting found and upon validating the credentials it reports 'Login Succeeded!' even though the registry does not exist and doesn't matter what is entered for user/pass.

Is there something I am missing here with using podman login?

**Comment by mheon** [comment]
@vrothberg @mtrmac PTAL

**Reply by vrothberg** [reply]
> Noticed this as a result of a typo, but it appears that when doing a podman login, as long as a curl GET returns success, that podman will report successful login regardless of the credentials entered. 

I am not sure there's something Podman can do.  If the server returns success, how could Podman verify it?

**Reply by gshepherd7** [reply]
Understand that portion, but what about the successful validation of the credentials?

**Reply by vrothberg** [reply]
I am not sure I understand the question. What should Podman do if the servers gives thumbs up?

**Reply by mtrmac** [reply]
When the server is not specifying any authentication mechanism = providing no way for credentials to be rejected, we assume things are OK.

It might be worth considering to explicitly fail if the user specifies credentials and the server doesn’t ask for any — OTOH by now that could well break production users. It’s also not quite clear that this would be appropriate — I could imagine a server that accepts authentication _either_ through a TLS certificate _or_ using an account:password credential pair, and it would be awkward for the user to have to log out before being allowed to use the TLS certificate.

**Reply by mtrmac** [reply]
… I think it’s very unclear whether should reject _image operations_ when given credentials and the server doesn’t ask for any.

The case for `docker.CheckAuth` failing in that situation is not 100% clear, but significantly stronger. (I’m not sure whether we can do that in the existing function or whether it would be better to add a new one, but that’s a fairly trivial matter.)

**Reply by gshepherd7** [reply]
Understood.  This caused panic on our end from a copy & paste of the incorrect registry name when attempting to pull images, after the successful login response.  They thought their images were gone.  Wasn't sure if there was a way to handle this.  Is there a way I can check if the server does not require credentials prior to making the podman login request?

**Reply by mtrmac** [reply]
Parse the `WWW-Authenticate` header on the `/v2/` response.

Let’s see what other think about making “verify credentials” operation used for login to do that automatically.

**Reply by vrothberg** [reply]
I am reluctant to change the current behavior as we'd likely break production users as @mtrmac has mentioned earlier.

We could, however, add a `--verify` flag to make sure the credentials were passed to the registry?

**Reply by mtrmac** [reply]
For `podman login`, I am fairly confident that failing with an error for “you don’t need to log in to this registry, double-check that what you are doing” is appropriate — it’s 100% appropriate for interactive use, and probably for non-interactive use as well.

For a plain `podman pull`, if there are credentials in `auth.json`, I’m fairly strongly leaning towards _not_ failing.

For `docker.CheckAuth`, I don’t know — but we can introduce a new function easily enough, bypassing that question.

**Reply by vrothberg** [reply]
> For podman login, I am fairly confident that failing with an error for “you don’t need to log in to this registry, double-check that what you are doing” is appropriate — it’s 100% appropriate for interactive use, and probably for non-interactive use as well.

That sounds attractive. I trust your judgement.

> For a plain podman pull, if there are credentials in auth.json, I’m fairly strongly leaning towards not failing.

Totally agree on that.
